compiler-rt/lib/builtins/cpu_model/AArch64CPUFeatures.inc

```diff
@@ -73,8 +73,6 @@ enum CPUFeatures {
   FEAT_SSBS,
   FEAT_SSBS2,
   FEAT_BTI,
-  FEAT_LS64,
-  FEAT_LS64_V,
```

Comment:
I think we might need to keep the enum values stable, although I'm not sure whether this could actually cause any issues yet.
(c.f. https://github.com/llvm/llvm-project/pull/88965#pullrequestreview-2004684899, which didn't reach any conclusion)

Reply:
We **absolutely** must keep the values the same since they are now ABI and used in the ifunc resolvers in the object files:

```
        bl      __init_cpu_features_resolver
        adrp    x0, __aarch64_cpu_features
        ldr     x0, [x0, #:lo12:__aarch64_cpu_features]
        tbnz    x0, 36, .L8
```

The 36 is from the enum, now imagine a new compiler release changing it to 35 or a different compiler using 40...

Reply:
Thanks for flagging this up. I am puzzled though since I was under the impression that FMV is still not production ready. When did the enum become part of the ABI? I guess a workaround is to keep a reserved position in the enum when removing a feature.

Reply:
https://github.com/llvm/llvm-project/pull/88965#discussion_r1567997481

Reply:
https://github.com/llvm/llvm-project/pull/88965#discussion_r1567997481

> Did any implementation ship __aarch64_cpu_features in a way that would make this an ABI break? We (Apple) talked about putting these in libSystem, and now I am glad we did not. It's okay to change/reorder them if it is a private implementation detail between the compiler and compiler-rt that is shipped with it, but as soon as they're part of an ABI, this list must become append only.

https://github.com/llvm/llvm-project/pull/88965#discussion_r1568423939

> For the time being __aarch64_cpu_features is not part of the ABI, but that may change in the future.

https://github.com/llvm/llvm-project/pull/88965#discussion_r1570826110

> For the ifunc implementation to work reliably, then the relevant parts of compiler-rt (or libgcc for the GCC implementation) should be statically linked. Otherwise there's a possibility that the symbols aren't actually loaded yet when the ifunc resolver is run. So there should be no exposed ABI in the context of FMV.

> As for other uses, we haven't agreed to freeze this as part of the ABI yet, but we are giving this possibility some consideration in designing the FMV resolver.

Reply:
It became a defacto ABI when the first implementation became available. We could change the names of the global/init function to explicitly break the ABI with old objects at linktime to avoid silent ABI breaks that lead to incorrect execution at runtime.

Reply:
We could do that automatically every release by putting the version number in its name.

Reply:
Yes, that's a way to keep backwards compatibility with older ABIs. However we can't break this ABI every compiler release (except perhaps this one time).

Reply:
@Wilco1 could you explain the scenario you are worried about in more detail? What you posted above appears to be a snippet of a generated resolver, suggesting that the problematic scenario is:
- You have an old object file which contains a resolver.
- The resolver refers to bit 36 and has references to `__init_cpu_features_resolver` (defined in `compiler-rt`)
- You link against a new `compiler-rt`, which has changed the meaning of bit 36.

I don't think step 3 is allowed. The compiler and compiler-rt need to remain in sync, afaik?

There are no references to specific bit numbers outside of the generated resolvers and `__init_cpu_features_resolver`.

Reply:
ABI compatibility means you can use different compilers to build object files, link them and they work correctly. If object A uses bit 36, object B uses bit 35 and the runtime uses bit 34, how is it ever going to work?

So using an enum in the implementation is really a bug. The FEAT_xxx bits are defined as constants for a good reason. And yes, these are ABI too!

Reply:
>  If object A uses bit 36, object B uses bit 35 and the runtime uses bit 34, how is it ever going to work?

Using them for what? Where are these bits being passed across an object file boundary?

Reply:
They are stored in the global `__aarch64_cpu_features` which is set by the init code and used by every FMV ifunc resolver.